pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -230,248 +259,327 @@ func (r *ReconcileBareMetalHost) Reconcile(request reconcile.Request) (reconcile
 		if err != nil {
-			return reconcile.Result{}, errors.Wrap(err, "failed to validate BMC access")
+			return reconcile.Result{}, errors.Wrap(err, fmt.Sprintf("phase %s failed", phase.name))
+		}
+		if phaseResult == nil {
+			continue
 		}
 
-		// We may have changed the host during validation or when we
-		// cleared the error state above. In either case, save now.
-		reqLogger.Info("after validation", "provResult", provResult)
-		if provResult.Dirty {
-			reqLogger.Info("saving status after validating management access")
-			if err := r.saveStatus(host); err != nil {
+		if phaseResult.Requeue {
+			ctx.log.Info("saving host status")
+			if err = r.saveStatus(host); err != nil {
 				return reconcile.Result{}, errors.Wrap(err,
-					"failed to save provisioning host status")
-			}
-			// We need to wait before checking with the provisioner
-			// again.
-			reqLogger.Info("host not ready", "delay", provResult.RequeueAfter)
-			result := reconcile.Result{
-				Requeue:      true,
-				RequeueAfter: provResult.RequeueAfter,
+					fmt.Sprintf("failed to save host status after %s phase", phase.name))
 			}
-			return result, nil
 		}
 
 		if host.HasError() {
-			// We have tried to register and validate the host and
-			// that failed in a way we assume is not retryable, so do
-			// not proceed to any other steps.
-			reqLogger.Info("registration error")
+			// We have tried to do something that failed in a way we
+			// assume is not retryable, so do not proceed to any other
+			// steps.
+			ctx.log.Info("stopping on host error")
 			return reconcile.Result{}, nil
 		}
 
-		// Reaching this point means the credentials are valid and
-		// worked, so record that in the status block.
-		reqLogger.Info("updating credentials success status fields")
-		host.UpdateGoodCredentials(*bmcCredsSecret)
+		ctx.log.Info("phase done",
+			"requeue", (*phaseResult).Requeue,
+			"after", (*phaseResult).RequeueAfter,
+		)
+		return *phaseResult, nil
+	}
+
+	// If we have nothing else to do and there is no LastUpdated
+	// timestamp set, set one.
+	if host.Status.LastUpdated.IsZero() {
+		reqLogger.Info("initializing status")
 		if err := r.saveStatus(host); err != nil {
-			return reconcile.Result{}, errors.Wrap(err,
-				"failed to update credentials success status fields")
+			return reconcile.Result{}, errors.Wrap(err, "failed to initialize status block")
 		}
 	}
-	if host.Status.Provisioning.State == provisioner.StateRegistrationError {
-		// We have tried to register and validate the host and that
-		// failed, so do not proceed to any other steps.
-		reqLogger.Info("registration error, stopping")
-		return reconcile.Result{}, nil
+
+	// Come back in 60 seconds to keep an eye on the power state
+	reqLogger.Info("Done with reconcile")
+	return reconcile.Result{RequeueAfter: time.Second * 60}, nil
+}
+
+// Handle delete operations.
+func (r *ReconcileBareMetalHost) phaseDelete(ctx reconcileContext) (result *reconcile.Result, err error) {
```

Comment:
Same context issue as above. It's very common for the first argument of a function to be `func foo(ctx context.Context)`. This could be confusing. It's not as bad as naming the first parameter of a non-method python function `self`, but maybe 30% of the way there.